SU2_CFD/include/limiters/CLimiterDetails.hpp

```diff
@@ -0,0 +1,335 @@
+/*!
+ * \file CLimiterDetails.hpp
+ * \brief A template class that allows defining limiters via
+ *        specialization of particular details.
+ * \author P. Gomes
+ * \version 7.0.0 "Blackbird"
+ *
+ * SU2 Project Website: https://su2code.github.io
+ *
+ * The SU2 Project is maintained by the SU2 Foundation
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2019, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+/*!
+ * \brief A traits class for limiters, see notes for "computeLimiters_impl()".
+ * \note There is no default implementation (the code will compile but not
+ *       link) specialization is mandatory.
+ */
+template<ENUM_LIMITER LimiterKind>
+struct CLimiterDetails
+{
+  /*!
+   * \brief Compute any global value that may be needed by the other functions.
+   * \note This function is called once by multiple threads.
+   */
+  template<class FieldType>
+  inline void preprocess(CGeometry&, CConfig&, size_t varBegin,
+                         size_t varEnd, const FieldType&);
+
+  /*!
+   * \brief Geometric modifier (e.g. increase limiting near sharp edges).
+   * \note This function is called once per point inside an AD pre-
+   *       -accumulation region, newly used variables should be registered.
+   */
+  inline su2double geometricFactor(size_t iPoint, CGeometry&) const;
+
+  /*!
+   * \brief Smooth (usually) function of the maximum/minimum (positive/negative)
+   *        gradient projections onto the edges, and the deltas over direct neighbors.
+   *        Both proj and delta may be 0.0, beware of divisions.
+   * \note This function is called twice (min/max) per point per variable
+   *       (also inside an AD pre-accumulation region).
+   */
+  inline su2double limiterFunction(size_t iVar, su2double proj, su2double delta) const;
+};
+
+
+/*!
+ * \brief Common small functions used by limiters.
+ */
+namespace LimiterHelpers
+{
+  inline passivedouble epsilon() {return std::numeric_limits<passivedouble>::epsilon();}
+
+  inline su2double venkatFunction(su2double proj, su2double delta, su2double eps2)
+  {
+    su2double y = delta*(delta+proj) + eps2;
+    return (y + delta*proj) / (y + 2*proj*proj);
+  }
+
+  inline su2double raisedSine(su2double dist)
+  {
+    su2double factor = 0.5*(1.0+dist+sin(PI_NUMBER*dist)/PI_NUMBER);
+    return max(0.0, min(factor, 1.0));
+  }
```

Comment:
Initially I took this function from the sharp-edge limiter but I think it was incorrect (the way the distance was normalized and the bounds checking (replaced by min/max)  did not seem to be continuous).
Now for distances less than the reference the "geometric limiter" is 0, at 3 times the distance the value is 1.